### PR TITLE
ENH: improved error message `IndexError: too many indices for array`

### LINKED ISSUE #15321: ENH: Improve error message: `IndexError: too many indices for array`
**IndexError: too many indices for array**

### error example:

File "object_detection/mobilenet_test.py", line 153, in <module>
    line_thickness=8)
  File "/home/abhisek/Desktop/models/research/object_detection/utils/visualization_utils.py", line 798, in visualize_boxes_and_labels_on_image_array
    if scores is None or scores[i] > min_score_thresh:
IndexError: too many indices for array

> Unable to start the inference due to the error " IndexError: too many indices for array "

### PR DESCRIPTION
**Pull request by LaisaLI**
ENH: improved error message `IndexError: too many indices for array` when the dimension of the index is larger than the dimension of the array, fixes #15321

Author:  Yilin Li <yilin.lasia@gmail.com>

### CODE REVIEW

File: numpy/core/src/multiarray/mapping.c [hunk 1]
```diff
@@ -727,8 +727,11 @@ prepare_index(PyArrayObject *self, PyObject *index,
         }
     }
     else if (used_ndim > PyArray_NDIM(self)) {
-        PyErr_SetString(PyExc_IndexError,
-                        "too many indices for array");
+        PyErr_Format(PyExc_IndexError,
+                     "array has %"NPY_INTP_FMT
+                     "-dimension, but %d were indexed",
+                     PyArray_NDIM(self),
+                     used_ndim);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/mapping.c]: It would be good to keep `"too many"` in the message somewhere.

Comment (LaisaLI) [reply numpy/core/src/multiarray/mapping.c]: Could you give some example code?

Comment (seberg) [reply numpy/core/src/multiarray/mapping.c]: Note that `ndim` is an integer IIRC, so it should just use `%d`. I am not sure what you mean with example code? Eric is merely suggesting to start the message with something like the old message (maybe rearranged) before you go into the details.

File: numpy/core/tests/test_indexerrors.py [hunk 2]
```diff
@@ -1,5 +1,8 @@
 import numpy as np
-from numpy.testing import assert_raises
+from numpy.testing import (
+        assert_raises, assert_raises_regex
```
Comment (seberg) [inline numpy/core/tests/test_indexerrors.py]: Just nitpicking (don't bother, I constantly ignore it), but I think if the closing bracket is on the next line, it is good style to end with a comma.

File: numpy/core/src/multiarray/mapping.c (hunk 1 above)
Comment (LaisaLI) [reply numpy/core/src/multiarray/mapping.c]: I see! Sorry, I misunderstood that.

Comment (LaisaLI) [reply numpy/core/src/multiarray/mapping.c]: How about `too many indices for array: array has 2-dimension, but 5 were indexed`? Is it more clear?

File: numpy/core/tests/test_indexerrors.py (hunk 2 above)
Comment (LaisaLI) [reply numpy/core/tests/test_indexerrors.py]: Thanks for your advice! I'm really new to Numpy. I'd love to accept this kind of advice to keep the same code style with other code!

Comment (mattip) [reply numpy/core/tests/test_indexerrors.py]: The idea behind ending the line with a comma is that it allows additional modification in the future with minimal disruption to the code

File: numpy/core/src/multiarray/mapping.c [hunk 3]
```diff
@@ -727,8 +727,11 @@ prepare_index(PyArrayObject *self, PyObject *index,
         }
     }
     else if (used_ndim > PyArray_NDIM(self)) {
-        PyErr_SetString(PyExc_IndexError,
-                        "too many indices for array");
+        PyErr_Format(PyExc_IndexError,
+                     "too many indices for array: "
+                     "array has %d-dimension, but %d were indexed",
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/mapping.c]:
```suggestion
                     "array is %d-dimensional, but %d were indexed",
```

File: numpy/core/tests/test_indexerrors.py [hunk 4]
```diff
@@ -110,6 +113,15 @@ def assign(obj, ind, val):
         assert_raises(IndexError, lambda: a[(1, [0, 1])])
         assert_raises(IndexError, lambda: assign(a, (1, [0, 1]), 1))
 
+    def test_mapping_error_message(self):
+        a = np.zeros((3, 5))
+        index = (1, 2, 3, 4, 5)
+        assert_raises_regex(
+                IndexError,
+                "too many indices for array: "
+                "array has 2-dimension, but 5 were indexed",
```
Comment (eric-wieser) [inline numpy/core/tests/test_indexerrors.py]:
```suggestion
                "array is 2-dimensional, but 5 were indexed",
```

**Review by eric-wieser — APPROVED** [review]

**Comment by mattip** [comment]
Thanks @LaisaLI

**Pull request merged**